text/0000-introduce empty ranges.md

```diff
@@ -0,0 +1,425 @@
+
+Combining these two criteria, for a `Range` to not be *empty*, progress from `min` to `max` must be possible (no-progess is `false`), and the iterator that realizes this progress must produce *finite* values that lie within `[min, max)`.
+
+Using this test, `Range(0, 10, -1)` is *empty* and no longer *infinite*, but so is in fact `Range(0, inf, -1)` that is also considered *infinite* currently. While one can generally never produce a complete list of elements that would incrementally move from say 0 to +Inf, we can know in the case of `Range(0, inf, -1)` that the Range must be *empty* because of the incompatible sign of the step parameter relative to the signs of `min` and `max` (progress impossible).
+Another case in which a Range is now *empty* is when the bound equality `min == max` can be evaluated meaningfully, independent of whether `step` equals e.g. `0` or `+-Inf` such as for example `Range(10, 10, 0)`. In the case of equal and finite bounds, no value can exist that satisfies `[min, min)`. In case of infinite bound values, the range is also empty because one `inf` cannot be meaningfully tested for equality to another `inf` and even if it could, the second criterion would be violated in that a sequence of values would contain non-finite values.
+
+As already stated, the current `Range` implementation treats the occurrence of `+-Inf` or `NaN` parameters in *any* of the `min, max, step` parameters as sufficient condition for *infinite*. This RFC considers the opposite: the no-progress expression cannot be meaningfully evaluated if any one of the 3 tested parameters is a `NaN` value, and one can also not decide whether any .next() iterations which add NaN lie within the given numerical range, or, if `min` or `max` are `NaN`, what that range is in the first place. Any occurrence of `NaN` in the Range parameters therefore produces an *empty* Range. Similarly, while a `step` argument of `+-Inf` may "pass" the inspection by the *no-progress criterion*, no finite points within the mathematical range `[min, max)` can be computed, including when `min, max` themselves are `inf`. Therefore, such Ranges are *empty*, too. A `step` argument of `0` satisfies the *no-progress criterion* and produces an *empty* Range.
```

Comment:
You jump from the current implementation to your proposed definitions in this paragraph without clear distinction of what currently is and what will be. The purpose here seems to be to discuss how `min`, `max`, and/or `step` being `+Inf`, `-Inf` or `NaN` will now produce empty ranges, but that is not true under your own rules. `Range(0, Inf, 1)` is not empty.

You also restate what you have already stated about the situation when `min == max` producing an empty range -- "...no finite points within the mathematical range `[min, max)` can be computed, including when `min, max` themselves are `inf`." -- this restatement is distracting from your point here.

Reply:
Yes, this should be shortened and simplified.

Reply:
>but that is not true under your own rules. Range(0, Inf, 1) is not empty.

Range(0, Inf, 1) is empty since progress is possible and only finite values will be created (in the first 1.7976931348623157e+308 iterations)

Reply:
> No, Range(0, Inf, 1) is not empty under the proposed criteria/rules, since progress is possible and only finite values will be created (in the first 1.7976931348623157e+308 iterations). I think the sentence
Correct, I missed this distinction and think that was due mostly to me being tired, not to wording.